osu.Game/Beatmaps/WorkingBeatmap_VirtualBeatmapTrack.cs

```diff
@@ -0,0 +1,50 @@
+﻿// Copyright (c) 2007-2018 ppy Pty Ltd <contact@ppy.sh>.
+// Licensed under the MIT Licence - https://raw.githubusercontent.com/ppy/osu/master/LICENCE
+
+using System.Linq;
+using osu.Framework.Audio.Track;
+using osu.Game.Rulesets.Objects.Types;
+
+namespace osu.Game.Beatmaps
+{
+    public partial class WorkingBeatmap
+    {
+        /// <summary>
+        /// A type of <see cref="TrackVirtual"/> which provides a valid length based on the <see cref="HitObject"/>s of an <see cref="IBeatmap"/>.
+        /// </summary>
+        private class VirtualBeatmapTrack : TrackVirtual
+        {
+            private readonly IBeatmap beatmap;
+
+            public VirtualBeatmapTrack(IBeatmap beatmap)
+            {
+                this.beatmap = beatmap;
+                updateVirtualLength();
+            }
+
+            protected override void UpdateState()
+            {
+                updateVirtualLength();
```

Comment:
why is the update here required? shouldn't `beatmap` never change?

Reply:
Editor?

Reply:
i'd wait until we have the editor actually performing operations. we likely don't want this kind of thing running per-frame (but on a callback instead).